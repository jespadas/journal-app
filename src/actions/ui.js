import { types } from "../types/types";

// Type to set an Error it returns the error info
export const setError = (err) => ( {
    type: types.uiSetError,
    payload: err
} );

// Type to remove the Error message
export const removeError = () => ( {
    type: types.uiRemoveError
} );

// Type to set the loading spinner
export const startLoading = () => ( {
    type: types.uiStartLoading
} );

// Type to end the loading spinner
export const finishLoading = () => ( {
    type: types.uiFinishLoading
} );