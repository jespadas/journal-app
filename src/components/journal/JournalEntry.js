import React from "react";

export const JournalEntry = () => {
  return (
    <div className="journal__entry pointer">
      <div
        className="journal__entry-picture"
        style={{
          backgroundSize: "cover",
          backgroundImage:
            "url(https://cdn.pixabay.com/photo/2016/02/13/12/26/aurora-1197753_960_720.jpg)",
        }}
      ></div>
      <div className="journal__entry-body">
        <p className="journal__entry-title">Entrada nueva</p>
        <p className="journal__entry-content">Entrada nueva contenido</p>
      </div>
      <div className="journal__entry-date-box">
        <span>Monday</span>
        <h4>28</h4>
      </div>
    </div>
  );
};
