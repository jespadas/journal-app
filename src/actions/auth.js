import Swal from 'sweetalert2';

import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { types } from '../types/types';
import { noteLogout } from './notes';
import { finishLoading, startLoading } from './ui';

// Action to login with the Email and Passwoord
export const startLoginEmailPassword = (email, password) => {
	return (dispatch) => {
		// Dispatch the loader spinner
		dispatch(startLoading());
		// Connection into firebase and auth
		return firebase
			.auth()
			.signInWithEmailAndPassword(email, password)
			// If ok it dispatch the login action with user credentials
			.then(async ({ user }) => {
				await dispatch(login(user.uid, user.displayName));
				// Message
				Swal.fire('Success', 'Welcome !', 'success');
				// Dispatch the end of loader spinner
				dispatch(finishLoading());
			})
			// Catch the error, it ends the loader spinner and show
			// the error message
			.catch((e) => {
				dispatch(finishLoading());
				Swal.fire('Error', e.message, 'error');
			});
	};
};

// Action to login with form data
export const startRegisterWithEmailPasswordName = (email, password, name) => {
	return (dispatch) => {
		// Connection into firebase and auth
		firebase
			.auth()
			.createUserWithEmailAndPassword(email, password)
			// If ok it dispatch the login action with user credentials
			.then(async ({ user }) => {
				// Await the response from updating user profile action
				// Updates the user display name with the received name
				await user.updateProfile({ displayName: name });
				// Dispatch the new values
				dispatch(login(user.uid, user.displayName));
			})
			// Catch the error and show the message with its info
			.catch((e) => {
				Swal.fire('Error', e.message, 'error');
			});
	};
};

// Action to login with Google credentials
export const startGoogleLogin = () => {
	return (dispatch) => {
		// Connection into firebase and auth and show Popup
		firebase
			.auth()
			.signInWithPopup(googleAuthProvider)
			// If ok it dispatch the received user credentials
			.then(({ user }) => {
				dispatch(login(user.uid, user.displayName));
			})
			.catch((e) => {
				Swal.fire('Error', e.message, 'error');
			});
	};
};

// Sends the news values to reducer
export const login = (uid, displayName) => {
	return {
		type: types.login,
		payload: {
			uid,
			displayName,
		},
	};
};

// Action to logout
export const startLogout = () => {
	return async (dispatch) => {
		// Connection into firebase and logout
		await firebase
			.auth()
			.signOut()
			// If ok Dispatch the logout action type to
			// flush user credentials and the notes data
			.then(dispatch(logout()), dispatch(noteLogout()))
			.catch((e) => {
				Swal.fire('Error', e.message, 'error');
			});
		/* dispatch(logout());
		dispatch(noteLogout()); */
	};
};

// Action to logout
export const logout = () => ({
	type: types.logout,
});
