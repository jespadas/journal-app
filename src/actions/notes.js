import Swal from 'sweetalert2';

import { db } from '../firebase/firebase-config';
import { fileUpload } from '../helpers/fileUpload';
import { loadNotes } from '../helpers/loadNotes';
import { types } from '../types/types';

// Action to post a new note
export const startNewNote = () => {
	return async (dispatch, getState) => {
		// Get the user id credential
		const { uid } = getState().auth;
		// Set empty values in a new note
		const newNote = {
			title: '',
			body: '',
			date: new Date().getTime(),
		};
		// Post the new note in data base
		const docRef = await db.collection(`${uid}/journal/notes`).add(newNote);
		// Dispatch the active note in order to set active value
		dispatch(activeNote(docRef.id, newNote));
		// Dispatch
		dispatch(addNewNote(docRef.id, newNote));
	};
};

// Load the notes and dispatch the data thru the action
export const startLoadingNotes = (uid) => {
	return async (dispatch) => {
		const notes = await loadNotes(uid);
		dispatch(setNotes(notes));
	};
};

// Return the active note
export const activeNote = (id, note) => ({
	type: types.notesActive,
	payload: {
		id,
		...note,
	},
});

// Set the state values from the new note
export const addNewNote = (id, note) => ({
	type: types.notesAddNew,
	payload: {
		id,
		...note,
	},
});

// Set the selected note
export const setNotes = (notes) => ({
	type: types.notesLoad,
	payload: notes,
});

// Start the async action to save the note in data base
export const startSaveNote = (note) => {
	return async (dispatch, getState) => {
		// Get user id credential
		const { uid } = getState().auth;
		// If there is not note url it delete it
		if (!note.url) {
			delete note.url;
		}
		// Save the values in the variable
		const noteToFirestore = { ...note };
		// Deletes the current id
		delete noteToFirestore.id;
		// Updates the note with its id
		await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);
		// Dispatch new values
		dispatch(refreshNote(note.id, noteToFirestore));
		// Message
		Swal.fire('Saved', note.title, 'success');
	};
};

// Refresh only the note selected by passing the id
export const refreshNote = (id, note) => ({
	type: types.notesUpdated,
	payload: {
		id,
		// Passing all the note properties
		note: {
			id,
			...note,
		},
	},
});

// Upload the selected file
export const startUploading = (file) => {
	return async (dispatch, getState) => {
		// Get the active note data
		const { active: activeNote } = getState().notes;
		// Message
		Swal.fire({
			title: 'Uploading',
			text: 'Please wait...',
			allowOutsideClick: false,
			onBeforeOpen: () => {
				Swal.showLoading();
			},
		});
		// Save the uploaded file in variable
		const fileUrl = await fileUpload(file);
		// Updates de active note url value with the new one
		activeNote.url = fileUrl;
		// Dispatch the new value
		dispatch(startSaveNote(activeNote));
		// End of message
		Swal.close();
	};
};


// Start deleting action
export const startDeleting = (id) => {
	return async (dispatch, getState) => {
		// Message
		Swal.fire({
			title: 'Deleting',
			text: 'Please wait...',
			allowOutsideClick: false,
			onBeforeOpen: () => {
				Swal.showLoading();
			},
		});
		// Get the user id credential
		const uid = getState().auth.uid;
		// Deletes the note from data base
		await db.doc(`${uid}/journal/notes/${id}`).delete();
		// If everything is ok, delete the note from store
		dispatch(deleteNote(id));
		// End of message
		Swal.close();
	};
};

// It send the id from the note selected to delete
export const deleteNote = (id) => ({
	type: types.notesDelete,
	payload: id,
});

// Action to flush the notes data
export const noteLogout = () => ({
	type: types.notesLogoutCleaning,
});
