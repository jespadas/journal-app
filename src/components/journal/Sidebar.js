import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { startLogout } from '../../actions/auth';
import { startNewNote } from '../../actions/notes';
import { JournalEntries } from './JournalEntries';

export const Sidebar = () => {
	// Get the name from state.auth.name
	const { name } = useSelector((state) => state.auth);

	// Set dispatch to use it
	const dispatch = useDispatch();

	// Handles the logout onCLick
	const handleLogout = () => {
		// Dispatch the start of logout action
		dispatch(startLogout());
	};
	
	// Handles the new entry onClick
	const handleNewEntry = () => {
		// Dispatch the start of adding a new note action
		dispatch(startNewNote());
	};

	return (
		<aside className='journal__sidebar'>
			<div className='journal__sidebar-navbar'>
				<h3 className='mt-5'>
					<i className='far fa-moon'></i>
					<span> {name}</span>
				</h3>
				<button className='btn' onClick={handleLogout}>
					Logout
				</button>
			</div>
			<div className='journal__new-entry' onClick={handleNewEntry}>
				<i className='far fa-calendar-plus fa-5x'></i>
				<p className='mt-5'>New entry</p>
			</div>
			<JournalEntries />
		</aside>
	);
};
