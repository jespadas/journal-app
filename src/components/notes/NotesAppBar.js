import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import moment from 'moment';
import { startSaveNote, startUploading } from '../../actions/notes';

export const NotesAppBar = ({ date }) => {
	// Formatting the date
	const noteDate = moment(date);

	// Set dispatch to use it
	const dispatch = useDispatch();

	// Get the active state data from state.notes.active
	const { active } = useSelector((state) => state.notes);

	// Handles saving the note onClick action
	const handleSave = () => {
		// Dispatch the start of saving a note action with
		// The active note info
		dispatch(startSaveNote(active));
	};

	// Handles the selection window to pick a file
	const handlePicture = () => {
		document.querySelector('#fileSelector').click();
	};

	// Handles the selected file action
	const handleFile = (e) => {
		// Get the file data
		const file = e.target.files[0];
		// If there is a file it dispatch it to the
		// Start uploading action with the file data
		if (file) {
			dispatch(startUploading(file));
		}
	};

	return (
		<div className='notes__appbar'>
			<span>{noteDate.format('MMM / D / YYYY')}</span>
			<input
				id='fileSelector'
				type='file'
				name='file'
				style={{ display: 'none' }}
				onChange={handleFile}
			/>
			<div>
				<button className='btn' onClick={handlePicture}>
					Picture
				</button>
				<button className='btn' onClick={handleSave}>
					Save
				</button>
			</div>
		</div>
	);
};
