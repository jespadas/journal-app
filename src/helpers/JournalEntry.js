import React from 'react';
import { useDispatch } from 'react-redux';
import moment from 'moment';
import { activeNote } from '../actions/notes';

export const JournalEntry = ({ id, date, title, body, url }) => {
	const noteDate = moment(date);

	const dispatch = useDispatch();

	const handleEntryClick = () => {
		dispatch(activeNote(id, { date, title, body, url }));
	};

	return (
		<div className='journal__entry pointer' onClick={handleEntryClick}>
			{url && (
				<div
					className='journal__entry-picture'
					style={{
						backgroundSize: 'cover',
						backgroundImage: `url(${url})`,
					}}
				></div>
			)}
			<div className='journal__entry-body'>
				<p className='journal__entry-title'>{title}</p>
				<p className='journal__entry-content'>{body}</p>
			</div>
			<div className='journal__entry-date-box'>
				<span>{noteDate.format('MMM')}</span>
				<h4>{noteDate.format('D')}</h4>
				<span>{noteDate.format('YYYY')}</span>
			</div>
		</div>
	);
};
