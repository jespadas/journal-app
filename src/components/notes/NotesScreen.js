import React from 'react';
import { NotesAppBar } from './NotesAppBar';

export const NotesScreen = () => {
	return (
		<div className='notes__main-content'>
			<NotesAppBar />
			<div className='notes__content'>
				<input
					type='text'
					placeholder='Entry your title'
					className='notes__input-title'
					autoFocus={true}
				/>
				<textarea
					placeholder='Entry content'
					className='notes__textarea'
				></textarea>

				<div className='notes__image'>
					<img
						src='https://cdn.pixabay.com/photo/2016/02/13/12/26/aurora-1197753_960_720.jpg'
						alt=''
					/>
				</div>
			</div>
		</div>
	);
};
