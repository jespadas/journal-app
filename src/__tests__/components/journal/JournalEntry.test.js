import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import '@testing-library/jest-dom';

// It is a mock store for testing Redux async action creators and middlewares
import configureStore from 'redux-mock-store';
// React middleware for Redux.
import thunk from 'redux-thunk';
import { JournalEntry } from '../../../components/journal/JournalEntry';
import { activeNote } from '../../../actions/notes';

// Adding middlewares (thunk)
const middlewares = [thunk];
// Creating the mockStore with middlewares
const mockStore = configureStore(middlewares);
const initState = {};
// Store
let store = mockStore(initState);
// mocking store dispatch
store.dispatch = jest.fn();

const note = {
    id: 10,
    date: 12,
    title: "Test title",
    body: "Test body",
    url: "https://test.com/test.png"
}

const wrapper = mount(
	<Provider store={store}>
		<JournalEntry {...note}/>
	</Provider>
);

describe('Tests in <JournalEntry />', () => {
	
    test('should be displayed correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should activate the note', () => {
        wrapper.find('.journal__entry').prop('onClick')();

        expect(store.dispatch).toHaveBeenCalled();
        expect(store.dispatch).toHaveBeenCalledWith(activeNote(note.id, {...note}));

    })
    

});
