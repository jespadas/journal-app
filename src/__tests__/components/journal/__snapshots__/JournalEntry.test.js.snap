// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <JournalEntry /> should be displayed correctly 1`] = `
<div
  className="journal__entry pointer animate__animated animate__fadeIn animate__faster"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://test.com/test.png)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      Test title
    </p>
    <p
      className="journal__entry-content"
    >
      Test body
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
       
      January
       
    </span>
    <h4>
       
      01
       
    </h4>
    <span>
       
      1970
       
    </span>
  </div>
</div>
`;
